File: android/app/src/main/java/com/ethosprotocol/api/ApiClient.kt
package com.ethosprotocol.api

import android.util.Log
import com.ethosprotocol.BuildConfig
import com.ethosprotocol.models.*
import com.ethosprotocol.models.TwoFactorStatus
import com.ethosprotocol.models.Enable2FARequest
import com.ethosprotocol.models.Enable2FAResponse
import com.ethosprotocol.models.Verify2FARequest
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.engine.HttpClientEngine
import io.ktor.client.engine.android.*
import io.ktor.client.plugins.HttpRequestTimeoutException
import io.ktor.client.plugins.HttpTimeout
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.plugins.logging.*
import io.ktor.client.plugins.websocket.WebSockets
import io.ktor.client.request.*
import io.ktor.http.*
import io.ktor.serialization.kotlinx.json.*
import kotlinx.serialization.json.Json
import java.io.IOException
import java.security.SecureRandom
import javax.inject.Inject
import javax.inject.Singleton
import javax.net.ssl.SSLContext
import javax.net.ssl.TrustManager
import javax.net.ssl.TrustManagerFactory
import javax.net.ssl.X509TrustManager

sealed class ApiResult<out T> {
    // cachedAt is non-null only when data was served from OfflineCache rather than fetched
    // live, so callers can surface how stale it is (e.g. VaultViewModel/OfflineBanner).
    data class Success<T>(val data: T, val cachedAt: Long? = null) : ApiResult<T>()
    data class Error(val message: String, val code: Int = 0) : ApiResult<Nothing>()
    object NetworkUnavailable : ApiResult<Nothing>()
}

@Singleton
class ApiClient(
    private val tokenProvider: TokenProvider,
    private val networkMonitor: NetworkMonitor,
    private val offlineCache: OfflineCache,
    private val baseUrl: String,
    // Overridable so tests can substitute a MockEngine instead of hitting real Android
    // networking; production callers (AppModule) get the real Android engine, pre-configured
    // with certificate pinning, for free.
    //
    // #117: Certificate / public-key pinning for api.ethos-protocol.app. PinningTrustManager
    // wraps the system TrustManager and additionally verifies that at least one certificate
    // in the chain matches a pinned SPKI SHA-256 hash. See CertificatePinning.kt for the
    // rotation strategy.
    //
    // This has to be configured here, on the engine instance itself, rather than via
    // HttpClient(engine) { engine { sslManager = ... } } below: that `engine { }` builder
    // DSL is only available when HttpClient is constructed from an engine *factory*
    // (HttpClient(Android) { ... }), not from an already-built HttpClientEngine instance —
    // which is what's injected here for testability.
    engine: HttpClientEngine = Android.create {
        sslManager = { httpsURLConnection ->
            val systemTm = getSystemTrustManager()
            if (systemTm != null) {
                val pinner = CertificatePinner()
                val pinningTm = PinningTrustManager(pinner, systemTm)
                val sslContext = SSLContext.getInstance("TLS")
                sslContext.init(null, arrayOf<TrustManager>(pinningTm), null)
                httpsURLConnection.sslSocketFactory = sslContext.socketFactory
            }
        }
    },
    private val retryPolicy: RetryPolicy = RetryPolicy.networkDefault
) {
    companion object {
        private const val TAG = "ApiClient"
    }

    // internal (not private): VaultEventSocket reuses this same client/connection pool
    // to open the `/ws` connection documented in shared/api-contract.md, rather than
    // standing up a second HttpClient.
    internal val client = HttpClient(engine) {
        install(ContentNegotiation) {
            json(Json { ignoreUnknownKeys = true; isLenient = true })
        }
        install(Logging) {
            // Logging Redaction Policy (#111) — see shared/api-contract.md §Logging Redaction Policy.
            // Full request/response bodies (bearer token, 2FA secrets, vault balances, beneficiary
            // addresses, acceptance tokens) must never be written to logcat in any build.
            // LogLevel.INFO logs only HTTP method + URL + status — no body, no sensitive headers.
            // LogLevel.NONE in release ensures zero leakage even if a future log level change
            // is accidentally introduced in debug code that ships to release.
            level = if (BuildConfig.DEBUG) LogLevel.INFO else LogLevel.NONE
        }
        // No timeouts were configured previously, so a stalled connection (e.g. dead wifi
        // captive portal) could hang a request — and the caller's loading state — forever.
        install(HttpTimeout) {
            requestTimeoutMillis = 30_000
            connectTimeoutMillis = 15_000
            socketTimeoutMillis = 30_000
        }
        install(WebSockets)
    }

    // Derives the `wss://.../ws?vault_id={id}` URL from [baseUrl] (documented in
    // shared/api-contract.md) for VaultEventSocket.
    internal fun webSocketUrl(vaultId: String): String {
        val wsBase = baseUrl.replaceFirst("https://", "wss://").replaceFirst("http://", "ws://")
        return "$wsBase/ws?vault_id=$vaultId"
    }

    // Auth
    suspend fun getChallenge(): ApiResult<AuthChallenge> = get("/auth/challenge")
    suspend fun verifyPasskey(req: PasskeyVerifyRequest): ApiResult<AuthToken> = post("/auth/verify", req)
    suspend fun registerPasskey(req: PasskeyRegisterRequest): ApiResult<AuthToken> = post("/auth/register", req)
    suspend fun refreshToken(): ApiResult<AuthToken> = post("/auth/refresh", Unit, skipTokenRefresh = true)

    // Account recovery ("lost your device?") — shared contract with iOS's #5.
    suspend fun initiateRecovery(req: RecoveryInitiateRequest): ApiResult<RecoveryInitiateResponse> =
        post("/auth/recovery/initiate", req)
    suspend fun completeRecovery(req: RecoveryCompleteRequest): ApiResult<Unit> =
        post("/auth/recovery/complete", req)

    // Adds a passkey to the *currently authenticated* account (#207), distinct from
    // registerPasskey (new account) and completeRecovery (recovery for a signed-out user).
    suspend fun addPasskey(req: AddPasskeyRequest): ApiResult<PasskeyCredential> =
        post("/auth/credentials", req)

    // Passkey credential management (#206) — an account is not limited to a single passkey,
    // so listCredentials() always returns a list.
    suspend fun listCredentials(): ApiResult<List<PasskeyCredential>> = get("/auth/credentials")

    suspend fun revokeCredential(credentialId: String): ApiResult<Unit> =
        delete("/auth/credentials/$credentialId", Unit)

    // Vaults
    suspend fun listVaults(): ApiResult<List<Vault>> = get("/vaults")

    /**
     * Paginated variant of listVaults (#112).
     * Pass [after] = page.nextCursor to fetch subsequent pages.
     * Continue while [VaultPage.hasMore] == true.
     */
    suspend fun listVaults(limit: Int = 20, after: String? = null): ApiResult<VaultPage> {
        val path = buildString {
            append("/vaults?limit=$limit")
            if (after != null) append("&after=$after")
        }
        return get(path)
    }
    suspend fun getVault(id: String): ApiResult<Vault> = get("/vaults/$id")
    suspend fun createVault(req: CreateVaultRequest): ApiResult<Vault> = post("/vaults", req)
    suspend fun checkIn(vaultId: String): ApiResult<Unit> = post("/vaults/$vaultId/checkin", Unit)
    suspend fun deposit(vaultId: String, amount: Long): ApiResult<Vault> =
        post("/vaults/$vaultId/deposit", mapOf("amount" to amount))
    suspend fun withdraw(vaultId: String, amount: Long): ApiResult<Vault> =
        post("/vaults/$vaultId/withdraw", mapOf("amount" to amount))

    // Beneficiary
    // token: the acceptance token parsed from the /accept deep-link URL
    // (e.g. https://ethos-protocol.app/vaults/{id}/accept?token=<token>).
    // Required by the server — see shared/api-contract.md §POST /vaults/{id}/accept.
    suspend fun acceptBeneficiary(vaultId: String, token: String): ApiResult<Unit> =
        post("/vaults/$vaultId/accept", BeneficiaryAcceptRequest(vaultId = vaultId, token = token))
    suspend fun updateBeneficiary(vaultId: String, newBeneficiary: String): ApiResult<Vault> =
        post("/vaults/$vaultId/beneficiary", BeneficiaryUpdateRequest(newBeneficiary))

    // #218: sets or clears (via label = null) a vault's display label.
    suspend fun updateVaultLabel(vaultId: String, label: String?): ApiResult<Vault> =
        post("/vaults/$vaultId/label", VaultLabelUpdateRequest(label))

    // #217: paginated vault activity history. Mirrors listVaults(limit, after)'s
    // own cursor convention for consistency within this client.
    suspend fun getVaultHistory(vaultId: String, limit: Int = 20, after: String? = null): ApiResult<VaultHistoryPage> {
        val path = buildString {
            append("/vaults/$vaultId/history?limit=$limit")
            if (after != null) append("&after=$after")
        }
        return get(path)
    }

    // 2FA
    suspend fun get2FAStatus(vaultId: String): ApiResult<TwoFactorStatus> = get("/vaults/$vaultId/2fa/status")
    suspend fun enable2FA(vaultId: String, req: Enable2FARequest): ApiResult<Enable2FAResponse> =
        post("/vaults/$vaultId/2fa/enable", req)
    suspend fun verify2FA(vaultId: String, req: Verify2FARequest): ApiResult<Unit> =
        post("/vaults/$vaultId/2fa/verify", req)
    suspend fun disable2FA(vaultId: String): ApiResult<Unit> =
        post("/vaults/$vaultId/2fa/disable", Unit)
    suspend fun challenge2FA(vaultId: String): ApiResult<TwoFactorStatus> =
        post("/vaults/$vaultId/2fa/challenge", Unit)

    // Push
    suspend fun registerPushToken(token: String): ApiResult<Unit> =
        post("/notifications/register", PushRegistration(token = token))
    suspend fun unregisterPushToken(token: String): ApiResult<Unit> =
        delete("/notifications/register", PushRegistration(token = token))

    // Internals
    private suspend inline fun <reified T> get(path: String): ApiResult<T> {
        ensureFreshToken()
        if (!networkMonitor.isConnected) {
            val cached = offlineCache.load(path)
            return if (cached != null) ApiResult.Success(Json.decodeFromString(cached.data), cachedAt = cached.timestamp)
            else ApiResult.NetworkUnavailable
        }
        return runCatching {
            val response = withRetry(retryPolicy, ::isRetryableNetworkError) {
                client.get("$baseUrl$path") { bearerAuth() }
            }
            when (response.status.value) {
                in 200..299 -> {
                    val body: T = response.body()
                    offlineCache.save(path, Json.encodeToString(kotlinx.serialization.serializer(), body))
                    ApiResult.Success(body)
                }
                // The token the server rejected is no longer valid — clear it locally so it
                // isn't kept being sent, and so the UI correctly routes back to AuthScreen.
                401 -> { tokenProvider.clear(); ApiResult.Error("Unauthorized", 401) }
                404 -> ApiResult.Error("Not found", 404)
                else -> ApiResult.Error("Server error ${response.status.value}", response.status.value)
            }
        }.getOrElse { e -> ApiErrorMapper.toApiResult(e) { if (BuildConfig.DEBUG) Log.w(TAG, "$path failed", it) } }
    }

    private suspend inline fun <reified B, reified T> post(
        path: String,
        body: B,
        skipTokenRefresh: Boolean = false
    ): ApiResult<T> {
        if (!skipTokenRefresh) ensureFreshToken()
        if (!networkMonitor.isConnected) return ApiResult.NetworkUnavailable
        return runCatching {
            val response = client.post("$baseUrl$path") {
                bearerAuth()
                antiReplayHeaders()
                contentType(ContentType.Application.Json)
                setBody(body)
            }
            when (response.status.value) {
                in 200..299 -> ApiResult.Success(if (T::class == Unit::class) Unit as T else response.body())
                401 -> { tokenProvider.clear(); ApiResult.Error("Unauthorized", 401) }
                else -> ApiResult.Error("Server error ${response.status.value}", response.status.value)
            }
        }.getOrElse { e -> ApiErrorMapper.toApiResult(e) { if (BuildConfig.DEBUG) Log.w(TAG, "$path failed", it) } }
    }

    private suspend inline fun <reified B, reified T> delete(path: String, body: B): ApiResult<T> {
        ensureFreshToken()
        if (!networkMonitor.isConnected) return ApiResult.NetworkUnavailable
        return runCatching {
            val response = client.delete("$baseUrl$path") {
                bearerAuth()
                antiReplayHeaders()
                contentType(ContentType.Application.Json)
                setBody(body)
            }
            // Ktor does not throw on non-2xx responses by default, so the status must be
            // checked explicitly here (as get()/post() already do) — otherwise a failed
            // deletion (401/500/etc.) is silently reported back to callers as success.
            when (response.status.value) {
                in 200..299 -> ApiResult.Success(if (T::class == Unit::class) Unit as T else response.body())
                401 -> { tokenProvider.clear(); ApiResult.Error("Unauthorized", 401) }
                else -> ApiResult.Error("Server error ${response.status.value}", response.status.value)
            }
        }.getOrElse { e -> ApiErrorMapper.toApiResult(e) { if (BuildConfig.DEBUG) Log.w(TAG, "$path failed", it) } }
    }

    // Best-effort: refreshes the stored token when it's near its expiry so the request
    // about to be made uses a live token instead of one about to be rejected with a 401.
    // A refresh failure just falls through and lets the actual request surface the error.
    private suspend fun ensureFreshToken() {
        if (tokenProvider.token == null || !tokenProvider.isNearExpiry()) return
        val result = refreshToken()
        if (result is ApiResult.Success) tokenProvider.setSession(result.data)
    }

    // GET is the only idempotent verb this client issues — retrying POST/DELETE
    // automatically could double-submit a mutation (check-in, withdrawal, 2FA
    // disable, ...), so only get() calls withRetry with this predicate.
    // HttpRequestTimeoutException is checked explicitly because it subclasses
    // CancellationException (so HttpTimeout can cooperate with coroutine
    // cancellation) rather than IOException.
    private fun isRetryableNetworkError(e: Throwable): Boolean =
        e is HttpRequestTimeoutException || e is IOException

    private fun HttpRequestBuilder.bearerAuth() {
        tokenProvider.token?.let { header(HttpHeaders.Authorization, "Bearer $it") }
    }

    // Anti-replay headers (task #121, see shared/api-contract.md).
    // Applied to every mutating request (POST / DELETE). GET requests are
    // idempotent and do not require replay protection.
    //
    // X-Nonce : 32 cryptographically-random bytes, hex-encoded. The server
    //           stores seen nonces and rejects any duplicate within the token's
    //           validity window, preventing a captured request from being
    //           replayed later.
    // X-Timestamp : current Unix epoch in seconds. The server rejects requests
    //               where |server_time − timestamp| > 300 s (5-minute window),
    //               limiting the replay window to that duration even if the
    //               nonce store is unavailable.
    private fun HttpRequestBuilder.antiReplayHeaders() {
        val nonce = ByteArray(32).also { SecureRandom().nextBytes(it) }
            .joinToString("") { "%02x".format(it) }
        val timestamp = System.currentTimeMillis() / 1_000L
        header("X-Nonce", nonce)
        header("X-Timestamp", timestamp.toString())
    }
}

// ── #117 helper ────────────────────────────────────────────────────────────────

/**
 * Returns the first [X509TrustManager] from the default [TrustManagerFactory],
 * or `null` if none is available. Used by [ApiClient] to provide a delegate for
 * [PinningTrustManager] so standard chain validation still runs before the pin check.
 */
internal fun getSystemTrustManager(): X509TrustManager? {
    val factory = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm())
    factory.init(null as java.security.KeyStore?)
    return factory.trustManagers.filterIsInstance<X509TrustManager>().firstOrNull()
}

File: ios/EthosProtocol/Sources/Services/APIClient.swift
import Foundation

enum APIError: LocalizedError {
    case unauthorized
    case notFound
    case serverError(String)
    case networkUnavailable
    case decodingFailed

    var errorDescription: String? {
        switch self {
        case .unauthorized:       return "Authentication required"
        case .notFound:           return "Resource not found"
        case .serverError(let m): return m
        case .networkUnavailable: return "No internet connection"
        case .decodingFailed:     return "We couldn't read the server's response"
        }
    }

    /// A concrete next step to show alongside errorDescription. Without this, a
    /// decode/server failure just tells the user something's wrong with no
    /// indication of whether retrying will help or they should reach out — see
    /// ErrorPresentation, which turns this (plus isRetryable/suggestsContactSupport
    /// below) into the "Try Again" / "Contact Support" affordance shown in-app.
    var recoverySuggestion: String? {
        switch self {
        case .unauthorized:       return "Sign in again to continue."
        case .notFound:           return nil
        case .serverError:        return "Try again. If the problem continues, contact support."
        case .networkUnavailable: return "Check your connection and try again."
        case .decodingFailed:     return "Try again. If the problem continues, contact support — this has been logged for us to investigate."
        }
    }

    /// Whether a "Try Again" affordance makes sense for this error.
    var isRetryable: Bool {
        switch self {
        case .serverError, .networkUnavailable, .decodingFailed: return true
        case .unauthorized, .notFound: return false
        }
    }

    /// Whether "Contact Support" should be offered — reserved for failures the
    /// user can't resolve by retrying alone (a decode failure means the client
    /// and server response shape have drifted; a persistent server error may be
    /// an outage), as opposed to ones with a clear, actionable client-side fix
    /// (sign in again, check your connection).
    var suggestsContactSupport: Bool {
        switch self {
        case .decodingFailed, .serverError: return true
        case .unauthorized, .notFound, .networkUnavailable: return false
        }
    }
}

// `public` here (and on listAllVaults() below): TTLWidget.swift calls
// APIClient.shared.listAllVaults() across a real module boundary in the SPM
// build (Package.swift declares TTLWidget as a separate target depending
// on the EthosProtocol target) — internal (the Swift default) is invisible
// outside the defining module. Other members stay internal since only
// listAllVaults() is called from outside this module.
public final class APIClient {
    public static let shared = APIClient()

    // `internal` (not `private`): VaultEventSocket derives the wss:// URL from
    // the same configured base URL instead of hardcoding a second copy of it.
    let baseURL: URL
    private let session: URLSession
    private let decoder: JSONDecoder
    private let retryPolicy: RetryPolicy

    private convenience init() {
        let urlString = Bundle.main.object(forInfoDictionaryKey: "API_BASE_URL") as? String
            ?? "https://api.ethos-protocol.app/v1"
        // #117: Create a URLSession backed by PinningDelegate.
        // The delegate enforces public-key pinning against the hash(es) listed in
        // Info.plist under `TLS_PUBLIC_KEY_PINS`. Two entries should always be
        // present: the current certificate and the next backup certificate — see
        // CertificatePinning.swift for the rotation strategy.
        let pinningDelegate = PinningDelegate()
        let session = URLSession(
            configuration: .default,
            delegate: pinningDelegate,
            delegateQueue: nil
        )
        self.init(baseURL: URL(string: urlString)!,
                  session: session,
                  retryPolicy: .networkDefault)
    }

    // `internal` (not `private`): lets tests construct an APIClient with a mock
    // URLSession / RetryPolicy via `@testable import`. `.shared` remains the only
    // production entry point.
    init(baseURL: URL, session: URLSession, retryPolicy: RetryPolicy = .networkDefault) {
        self.baseURL = baseURL
        self.session = session
        self.retryPolicy = retryPolicy
        decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        decoder.dateDecodingStrategy = .iso8601
    }

    // MARK: - Auth

    func getChallenge() async throws -> AuthChallenge {
        try await post(path: "/auth/challenge", body: EmptyBody())
    }

    func verifyPasskey(credentialID: String, clientDataJSON: String, signature: String) async throws -> AuthToken {
        let body = ["credential_id": credentialID,
                    "client_data_json": clientDataJSON,
                    "signature": signature]
        return try await post(path: "/auth/verify", body: body)
    }

    func registerPasskey(credentialID: String, attestationObject: String, clientDataJSON: String) async throws {
        // Field name is `attestation_object` per shared/api-contract.md — the backend
        // parses the COSE public key out of the attestation object itself.
        // (Legacy field name `public_key` is not accepted by the server.)
        let body = ["credential_id": credentialID,
                    "attestation_object": attestationObject,
                    "client_data_json": clientDataJSON]
        let _: EmptyBody = try await post(path: "/auth/register", body: body)
    }

    // MARK: - Vaults

    /// One page of `GET /vaults`. See shared/api-contract.md's "List Pagination"
    /// section for the query-parameter/response-header contract this implements.
    struct VaultPage {
        let vaults: [Vault]
        let nextCursor: String?
    }

    static let defaultVaultPageSize = 50
    private static let nextCursorHeader = "X-Next-Cursor"

    /// Fetches a single page of the caller's vaults, starting after `cursor`
    /// (`nil` requests the first page). Drives VaultListView's "Load More".
    func listVaults(cursor: String? = nil, limit: Int = APIClient.defaultVaultPageSize) async throws -> VaultPage {
        var req = request(path: "/vaults", queryItems: Self.vaultsQueryItems(cursor: cursor, limit: limit))
        req.httpMethod = "GET"
        let (data, response) = try await execute(req)
        let vaults: [Vault] = try decode(data, path: "/vaults")
        let nextCursor = Self.parseNextCursor(fromHeaderValue: response.value(forHTTPHeaderField: Self.nextCursorHeader))
        return VaultPage(vaults: vaults, nextCursor: nextCursor)
    }

    /// Builds the `GET /vaults` query items for `cursor`/`limit` — split out from
    /// listVaults() so the cursor/limit contract is unit-testable independent of
    /// the network layer.
    static func vaultsQueryItems(cursor: String?, limit: Int) -> [URLQueryItem] {
        var queryItems = [URLQueryItem(name: "limit", value: String(limit))]
        if let cursor, !cursor.isEmpty {
            queryItems.append(URLQueryItem(name: "cursor", value: cursor))
        }
        return queryItems
    }

    /// An empty `X-Next-Cursor` header means "no further page", same as a missing one.
    static func parseNextCursor(fromHeaderValue value: String?) -> String? {
        (value?.isEmpty ?? true) ? nil : value
    }

    /// Fetches every page and concatenates the results, for callers (background
    /// refresh, the TTL widget) that need the complete vault set to compute
    /// accurate TTL warnings rather than a single page at a time.
    public func listAllVaults() async throws -> [Vault] {
        var all: [Vault] = []
        var cursor: String?
        repeat {
            let page = try await listVaults(cursor: cursor)
            all.append(contentsOf: page.vaults)
            cursor = page.nextCursor
        } while cursor != nil
        return all
    }

    /// Paginated variant of listVaults (#112).
    /// Pass `after: page.nextCursor` to fetch subsequent pages until `page.hasMore == false`.
    func listVaults(limit: Int = 20, after cursor: String? = nil) async throws -> VaultPage {
        var path = "/vaults?limit=\(limit)"
        if let cursor = cursor { path += "&after=\(cursor)" }
        return try await get(path: path)
    }

    func getVault(id: String) async throws -> Vault {
        try await get(path: "/vaults/\(id)")
    }

    func createVault(beneficiary: String, checkInInterval: UInt64) async throws -> Vault {
        let body = CreateVaultRequest(beneficiary: beneficiary, checkInInterval: checkInInterval)
        return try await post(path: "/vaults", body: body)
    }

    func checkIn(vaultID: String) async throws {
        let _: EmptyBody = try await post(path: "/vaults/\(vaultID)/checkin", body: EmptyBody())
    }

    func deposit(vaultID: String, amount: Int64) async throws -> Vault {
        try await post(path: "/vaults/\(vaultID)/deposit", body: ["amount": amount])
    }

    func withdraw(vaultID: String, amount: Int64) async throws -> Vault {
        try await post(path: "/vaults/\(vaultID)/withdraw", body: ["amount": amount])
    }

    func updateBeneficiary(vaultID: String, newBeneficiary: String) async throws -> Vault {
        try await post(path: "/vaults/\(vaultID)/beneficiary", body: ["beneficiary": newBeneficiary])
    }

    func acceptBeneficiary(vaultID: String, token: String) async throws {
        let body = ["vault_id": vaultID, "token": token]
        let _: EmptyBody = try await post(path: "/vaults/\(vaultID)/accept", body: body)
    }

    func getTTL(vaultID: String) async throws -> UInt64 {
        let result: [String: UInt64] = try await get(path: "/vaults/\(vaultID)/ttl")
        return result["ttl_remaining"] ?? 0
    }

    // MARK: - 2FA

    func get2FAStatus(vaultID: String) async throws -> TwoFactorStatus {
        try await get(path: "/vaults/\(vaultID)/2fa/status")
    }

    func enable2FA(vaultID: String, method: TwoFactorMethod, phone: String? = nil, email: String? = nil) async throws -> Enable2FAResponse {
        let body = Enable2FARequest(method: method, phone: phone, email: email)
        return try await post(path: "/vaults/\(vaultID)/2fa/enable", body: body)
    }

    func verify2FA(vaultID: String, otp: String) async throws {
        let _: EmptyBody = try await post(path: "/vaults/\(vaultID)/2fa/verify", body: Verify2FARequest(otp: otp))
    }

    func disable2FA(vaultID: String) async throws {
        let _: EmptyBody = try await post(path: "/vaults/\(vaultID)/2fa/disable", body: EmptyBody())
    }

    func challenge2FA(vaultID: String) async throws -> TwoFactorStatus {
        try await post(path: "/vaults/\(vaultID)/2fa/challenge", body: EmptyBody())
    }

    func clear2FASession(vaultID: String) async throws {
        let _: EmptyBody = try await post(path: "/vaults/\(vaultID)/2fa/session/clear", body: EmptyBody())
    }

    // MARK: - Push Notifications

    func registerPushToken(_ token: String) async throws {
        let body = PushRegistration(token: token, platform: "ios")
        let _: EmptyBody = try await post(path: "/notifications/register", body: body)
    }

    func unregisterPushToken(_ token: String) async throws {
        var req = request(path: "/notifications/register")
        req.httpMethod = "DELETE"
        req.httpBody = try? JSONEncoder().encode(PushRegistration(token: token, platform: "ios"))
        // Anti-replay: DELETE is a mutation; apply nonce + timestamp (task #121).
        for (field, value) in Self.makeAntiReplayHeaders() {
            req.setValue(value, forHTTPHeaderField: field)
        }
        _ = try await execute(req)
    }

    // MARK: - Logging Redaction Audit (#111)
    //
    // iOS uses URLSession directly — there is no logging plugin or interceptor in this file.
    // No request or response body, header, or sensitive field is written to os_log, print,
    // NSLog, or any other diagnostic channel anywhere in APIClient.swift.
    //
    // Invariant: any future addition of a logging layer to this client MUST:
    //   1. Be guarded by #if DEBUG ... #endif (or equivalent) so it is stripped from
    //      release builds entirely.
    //   2. Log only HTTP method, path (no query strings bearing tokens), and status code —
    //      never request/response bodies, Authorization headers, 2FA secrets, vault balances,
    //      beneficiary/owner wallet addresses, or acceptance tokens.
    //
    // See shared/api-contract.md §Logging Redaction Policy (#111) for the authoritative
    // cross-platform policy.

    // MARK: - Private helpers

    private func get<T: Decodable>(path: String) async throws -> T {
        var req = request(path: path)
        req.httpMethod = "GET"
        let (data, _) = try await execute(req)
        return try decode(data, path: path)
    }

    private func post<B: Encodable, T: Decodable>(path: String, body: B) async throws -> T {
        var req = request(path: path)
        req.httpMethod = "POST"
        req.httpBody = try JSONEncoder().encode(body)
        // Anti-replay: add nonce + timestamp to every mutating request (task #121).
        for (field, value) in Self.makeAntiReplayHeaders() {
            req.setValue(value, forHTTPHeaderField: field)
        }
        let data = try await execute(req)
        return try decode(data)
    }

    private func request(path: String, queryItems: [URLQueryItem]? = nil) -> URLRequest {
        let url = baseURL.appendingPathComponent(path)
        var finalURL = url
        if let queryItems, !queryItems.isEmpty,
           var components = URLComponents(url: url, resolvingAgainstBaseURL: false) {
            components.queryItems = queryItems
            finalURL = components.url ?? url
        }
        var req = URLRequest(url: finalURL)
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token = KeychainService.shared.loadToken() {
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        return req
    }

    // Anti-replay headers (task #121, see shared/api-contract.md).
    // Applied to every mutating request (POST / DELETE). GET requests are
    // idempotent and do not require replay protection.
    //
    // X-Nonce : 32 cryptographically-random bytes from CryptoKit, hex-encoded.
    //           The server stores seen nonces and rejects any duplicate within
    //           the token's validity window.
    // X-Timestamp : current Unix epoch in seconds. The server rejects requests
    //               where |server_time − timestamp| > 300 s (5-minute window).
    static func makeAntiReplayHeaders() -> [String: String] {
        // CryptoKit guarantees OS-CSPRNG quality randomness (SecRandomCopyBytes underneath).
        let nonceBytes = (0..<32).map { _ in UInt8.random(in: 0...255) }
        let nonce = nonceBytes.map { String(format: "%02x", $0) }.joined()
        let timestamp = String(Int(Date().timeIntervalSince1970))
        return ["X-Nonce": nonce, "X-Timestamp": timestamp]
    }

    private func execute(_ request: URLRequest) async throws -> Data {
        // Falling back to a cached response for a mutating request (POST/DELETE — e.g.
        // check-in, withdraw, disable2FA) would make the app report success for an action
        // that never actually reached the server, which is unacceptable for this app.
        let isCacheableRead = request.httpMethod == "GET"

        guard NetworkMonitor.shared.isConnected else {
            if isCacheableRead, let cached = OfflineCache.shared.load(for: request.url?.absoluteString ?? "") {
                // No real HTTP response exists for a cache hit — synthesize a bare 200 with
                // no headers, so e.g. listVaults()'s pagination-cursor header lookup just
                // reports "no more pages" instead of crashing on a missing response.
                let syntheticResponse = HTTPURLResponse(url: request.url ?? baseURL, statusCode: 200, httpVersion: nil, headerFields: nil)!
                return (cached, syntheticResponse)
            }
            throw APIError.networkUnavailable
        }

        let (data, response): (Data, URLResponse)
        if Self.isRetryable(method: request.httpMethod) {
            (data, response) = try await withRetry(retryPolicy, isRetryable: Self.isTransientNetworkError) {
                try await session.data(for: request)
            }
        } else {
            (data, response) = try await session.data(for: request)
        }
        guard let http = response as? HTTPURLResponse else { throw APIError.serverError("Invalid response") }
        switch http.statusCode {
        case 200...299:
            if isCacheableRead {
                OfflineCache.shared.save(data, for: request.url?.absoluteString ?? "")
            }
            return (data, http)
        case 401:
            // The token the server rejected is no longer valid — drop it locally so we
            // don't keep sending it, and so a relaunch correctly shows the sign-in screen.
            KeychainService.shared.deleteToken()
            throw APIError.unauthorized
        case 404: throw APIError.notFound
        default:
            let msg = (try? JSONDecoder().decode([String: String].self, from: data))?["error"] ?? "Server error"
            throw APIError.serverError(msg)
        }
    }

    private func decode<T: Decodable>(_ data: Data, path: String) throws -> T {
        if T.self == EmptyBody.self { return EmptyBody() as! T }
        do { return try decoder.decode(T.self, from: data) }
        catch {
            DecodingFailureLogger.shared.log(path: path, expectedType: String(describing: T.self), responseBody: data)
            throw APIError.decodingFailed
        }
    }

    // MARK: - Retry

    // GET is the only idempotent verb this client issues — retrying POST/DELETE
    // automatically could double-submit a mutation (check-in, withdrawal, 2FA
    // disable, ...), so only GET requests are eligible for `withRetry`.
    static func isRetryable(method: String?) -> Bool { method == "GET" }

    private static func isTransientNetworkError(_ error: Error) -> Bool {
        guard let urlError = error as? URLError else { return false }
        switch urlError.code {
        case .timedOut, .networkConnectionLost, .cannotConnectToHost,
             .cannotFindHost, .dnsLookupFailed, .notConnectedToInternet, .dataNotAllowed:
            return true
        default:
            return false
        }
    }
}

private struct EmptyBody: Codable {}

private struct CreateVaultRequest: Encodable {
    let beneficiary: String
    let checkInInterval: UInt64

    enum CodingKeys: String, CodingKey {
        case beneficiary
        case checkInInterval = "check_in_interval"
    }
}
